### ISSUE TITLE
Simple Convolutional Neural Network using TensorFlow

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#12113]
Simple Convolutional Neural Network using TensorFlow<br>
<br>
https://ift.tt/3x8qJVW<br>
<br>
Lee Meng<br>
<br>

<div><div><div><div>
<div><pre><span></span><span>batch_size</span> <span>=</span> <span>16</span>
<span>patch_size</span> <span>=</span> <span>5</span>
<span>depth</span> <span>=</span> <span>16</span>
<span>num_hidden</span> <span>=</span> <span>64</span>

<span>graph</span> <span>=</span> <span>tf</span><span>.</span><span>Graph</span><span>()</span>

<span>with</span> <span>graph</span><span>.</span><span>as_default</span><span>():</span>

    <span># Input data.</span>
    <span>tf_train_dataset</span> <span>=</span> <span>tf</span><span>.</span><span>placeholder</span><span>(</span>
        <span>tf</span><span>.</span><span>float32</span><span>,</span> <span>shape</span><span>=</span><span>(</span><span>batch_size</span><span>,</span> <span>image_size</span><span>,</span> <span>image_size</span><span>,</span> <span>num_channels</span><span>))</span>
    <span>tf_train_labels</span> <span>=</span> <span>tf</span><span>.</span><span>placeholder</span><span>(</span>
        <span>tf</span><span>.</span><span>float32</span><span>,</span> <span>shape</span><span>=</span><span>(</span><span>batch_size</span><span>,</span> <span>num_labels</span><span>))</span>
    <span>tf_valid_dataset</span> <span>=</span> <span>tf</span><span>.</span><span>constant</span><span>(</span><span>valid_dataset</span><span>)</span>
    <span>tf_test_dataset</span> <span>=</span> <span>tf</span><span>.</span><span>constant</span><span>(</span><span>test_dataset</span><span>)</span>

    <span># Variables.</span>
    <span># When defining weights for a convoluational layer, use the notation</span>
    <span># [filter_size, filter_size, input_depth, output_depth]</span>
    <span>layer1_weights</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span>
        <span>tf</span><span>.</span><span>truncated_normal</span><span>(</span>
            <span>[</span><span>patch_size</span><span>,</span> <span>patch_size</span><span>,</span> <span>num_channels</span><span>,</span> <span>depth</span><span>],</span> <span>stddev</span><span>=</span><span>0.1</span><span>))</span>
    <span>layer1_biases</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>tf</span><span>.</span><span>zeros</span><span>([</span><span>depth</span><span>]))</span>
    
    <span># in this CNN, two convoluational layers happen to have the same depth.</span>
    <span># if we want, we can adjust them to be different like depth1, depth2</span>
    <span>layer2_weights</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span>
        <span>tf</span><span>.</span><span>truncated_normal</span><span>(</span>
            <span>[</span><span>patch_size</span><span>,</span> <span>patch_size</span><span>,</span> <span>depth</span><span>,</span> <span>depth</span><span>],</span> <span>stddev</span><span>=</span><span>0.1</span><span>))</span>
    <span>layer2_biases</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>tf</span><span>.</span><span>constant</span><span>(</span><span>1.0</span><span>,</span> <span>shape</span><span>=</span><span>[</span><span>depth</span><span>]))</span>
    
    <span># because we use stride = 2 and SAME padding, our new shape of first feature map C1</span>
    <span># will be (image_size // 2, image_size //2). and because we use 2 convolutional layers,</span>
    <span># the shape of second feature map C2 will be (image_size // 2 // 2, image_size // 2 // 2)</span>
    <span># = (image_size // 4, image_size // 4). and because we have depth == 16,</span>
    <span># the total neurons on C2 will be image_size // 4 * image_size // 4 * depth</span>
    <span>layer3_weights</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span>
        <span>tf</span><span>.</span><span>truncated_normal</span><span>(</span>
            <span>[</span><span>image_size</span> <span>//</span> <span>4</span> <span>*</span> <span>image_size</span> <span>//</span> <span>4</span> <span>*</span> <span>depth</span><span>,</span> <span>num_hidden</span><span>],</span>
            <span>stddev</span><span>=</span><span>0.1</span><span>))</span>
    <span>layer3_biases</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>tf</span><span>.</span><span>constant</span><span>(</span><span>1.0</span><span>,</span> <span>shape</span><span>=</span><span>[</span><span>num_hidden</span><span>]))</span>
    <span>layer4_weights</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span>
        <span>tf</span><span>.</span><span>truncated_normal</span><span>([</span><span>num_hidden</span><span>,</span> <span>num_labels</span><span>],</span> <span>stddev</span><span>=</span><span>0.1</span><span>))</span>
    <span>layer4_biases</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>tf</span><span>.</span><span>constant</span><span>(</span><span>1.0</span><span>,</span> <span>shape</span><span>=</span><span>[</span><span>num_labels</span><span>]))</span>

    <span># Model.</span>
    <span>def</span> <span>model</span><span>(</span><span>data</span><span>):</span>
        <span># this is where we set stride = 2 for both width and height and also SAME padding</span>
        <span># the third parameters in tf.nn.conv2d is to set stride for every dimension</span>
        <span># specified in the first parameter data's shape</span>
        <span>conv</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>conv2d</span><span>(</span><span>data</span><span>,</span> <span>layer1_weights</span><span>,</span> <span>[</span><span>1</span><span>,</span> <span>2</span><span>,</span> <span>2</span><span>,</span> <span>1</span><span>],</span> <span>padding</span><span>=</span><span>'SAME'</span><span>)</span>
        <span>hidden</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>relu</span><span>(</span><span>conv</span> <span>+</span> <span>layer1_biases</span><span>)</span>
        <span>conv</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>conv2d</span><span>(</span>
            <span>hidden</span><span>,</span> <span>layer2_weights</span><span>,</span> <span>[</span><span>1</span><span>,</span> <span>2</span><span>,</span> <span>2</span><span>,</span> <span>1</span><span>],</span> <span>padding</span><span>=</span><span>'SAME'</span><span>)</span>
        <span>hidden</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>relu</span><span>(</span><span>conv</span> <span>+</span> <span>layer2_biases</span><span>)</span>
        <span>shape</span> <span>=</span> <span>hidden</span><span>.</span><span>get_shape</span><span>()</span><span>.</span><span>as_list</span><span>()</span>
        
        <span># turn the C2 3D cube back to 2D matrix by shape (#data_points, #neurons)</span>
        <span>reshape</span> <span>=</span> <span>tf</span><span>.</span><span>reshape</span><span>(</span><span>hidden</span><span>,</span>
                             <span>[</span><span>shape</span><span>[</span><span>0</span><span>],</span> <span>shape</span><span>[</span><span>1</span><span>]</span> <span>*</span> <span>shape</span><span>[</span><span>2</span><span>]</span> <span>*</span> <span>shape</span><span>[</span><span>3</span><span>]])</span>
        <span>hidden</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>relu</span><span>(</span><span>tf</span><span>.</span><span>matmul</span><span>(</span><span>reshape</span><span>,</span> <span>layer3_weights</span><span>)</span> <span>+</span> <span>layer3_biases</span><span>)</span>
        <span>return</span> <span>tf</span><span>.</span><span>matmul</span><span>(</span><span>hidden</span><span>,</span> <span>layer4_weights</span><span>)</span> <span>+</span> <span>layer4_biases</span>

    <span># Training computation.</span>
    <span>logits</span> <span>=</span> <span>model</span><span>(</span><span>tf_train_dataset</span><span>)</span>
    <span>loss</span> <span>=</span> <span>tf</span><span>.</span><span>reduce_mean</span><span>(</span>
        <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softmax_cross_entropy_with_logits</span><span>(</span>
            <span>labels</span><span>=</span><span>tf_train_labels</span><span>,</span> <span>logits</span><span>=</span><span>logits</span><span>))</span>

    <span># Optimizer.</span>
    <span>optimizer</span> <span>=</span> <span>tf</span><span>.</span><span>train</span><span>.</span><span>GradientDescentOptimizer</span><span>(</span><span>0.05</span><span>)</span><span>.</span><span>minimize</span><span>(</span><span>loss</span><span>)</span>

    <span># Predictions for the training, validation, and test data.</span>
    <span>train_prediction</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softmax</span><span>(</span><span>logits</span><span>)</span>
    <span>valid_prediction</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softmax</span><span>(</span><span>model</span><span>(</span><span>tf_valid_dataset</span><span>))</span>
    <span>test_prediction</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softmax</span><span>(</span><span>model</span><span>(</span><span>tf_test_dataset</span><span>))</span>
</pre></div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via LeeMeng - All About DataScience <br>
<br>
April 21, 2025 at 10:51AM
